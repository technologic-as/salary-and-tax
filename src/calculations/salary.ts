import { translationConfig } from '../translation';
import { getPension, pensionConstants } from './pension';
import round from './round';
import { ParamProps } from '../constants';

const hashParams = (defaults: ParamProps) => {
  try {
    const hash = window.location.hash.slice(1) || JSON.stringify(defaults);
    return JSON.parse(decodeURI(hash));
  } catch (e) {
    return defaults;
  }
};

export const defaultSalaryParameters = hashParams({
  cut: 10,
  hoursPerYear: 1730,
  hourRate: 1100,
  ensuringFee: 15000,
  vacationSavings: { rate: 12, include: true },
  employerFee: { rate: 14.1, include: true },
  surplus: { amount: 0, include: true, sevenG: true },
  graph: { increments: 10000 },
  locale: translationConfig.locale,
  ...pensionConstants,
});

function toNumber(string: number | string) {
  return parseFloat(`${string}`);
}

export const getTurnover = ({
  hoursPerYear,
  hourRate,
}: Pick<ParamProps, 'hoursPerYear' | 'hourRate'>): { turnover: number } => ({
  turnover: toNumber(hoursPerYear) * toNumber(hourRate),
});
export const getCompanyIncome = (
  turnover: number,
  { cut }: Pick<ParamProps, 'cut'>
): { after: number; theCut: number } => {
  const theCut = (turnover * toNumber(cut)) / 100;
  return { theCut: theCut, after: turnover - theCut };
};
export const calculateSurplus = (
  companyIncome: number,
  { surplus: { amount, include } }: Pick<ParamProps, 'surplus'>
) => {
  const includedAmount = include ? toNumber(amount) : 0;
  return { surplus: includedAmount, after: companyIncome - includedAmount };
};
export const getEmployerFee = (salary: number, { employerFee: { include, rate } }: Pick<ParamProps, 'employerFee'>) => {
  const employerFee = round(salary - (salary / (100 + (include ? toNumber(rate) : 0))) * 100);
  return { employerFee, after: salary - employerFee };
};
export const getVacationSavings = (
  salary: number,
  { vacationSavings: { include, rate } }: Pick<ParamProps, 'vacationSavings'>
) => {
  const vacationSavings = round(salary - (salary / (100 + (include ? toNumber(rate) : 0))) * 100);
  return { vacationSavings, after: salary - vacationSavings };
};

export type SalaryCalculationsState = {
  income: number;
  surplus: number;
  companyIncome: number;
  withoutVacationSavings: number;
  employerFee: number;
  theCut: number;
  vacationSavings: number;
  withoutEmployerFee: number;
  withoutSurplus: number;
  withoutPension: number;
  pension: number;
  totalCuts: number;
  turnover: number;
};
export const getSalaryCalculations = (data: ParamProps): SalaryCalculationsState => {
  const { turnover } = getTurnover(data);
  const { theCut, after: companyIncome } = getCompanyIncome(turnover, data);
  const { surplus, after: withoutSurplus } = calculateSurplus(companyIncome, data);
  const { employerFee, after: withoutEmployerFee } = getEmployerFee(withoutSurplus, data);
  const { vacationSavings, after: withoutVacationSavings } = getVacationSavings(withoutEmployerFee, data);
  const { pension, after: withoutPension } = getPension(withoutVacationSavings, data);
  const totalCuts = theCut + employerFee + vacationSavings + pension;

  return {
    turnover,
    theCut,
    companyIncome,
    employerFee,
    withoutEmployerFee,
    vacationSavings,
    withoutVacationSavings,
    pension,
    withoutPension,
    income: withoutPension,
    totalCuts,
    surplus,
    withoutSurplus,
  };
};
