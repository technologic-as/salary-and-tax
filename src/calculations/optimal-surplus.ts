import { calculateG } from './calculate-g';
import { calculate } from './index';
import { defaultSalaryParameters } from './salary';
import { ParamProps } from '../constants';

export type OptimalSurplus = {
  total: number;
  surplus: any;
  salaryBasis: number;
  dividends: number;
  netIncome: number;
  grossIncome: number;
  salaryAboveSevenG: boolean;
};

export const optimalSurplus = (parameters: ParamProps = defaultSalaryParameters): OptimalSurplus[] => {
  const initialCalculation = calculate(parameters);
  const companyIncome = initialCalculation.salary.companyIncome;
  const increments = parameters.graph.increments;
  const steps = Math.round(companyIncome / increments);

  const sevenG = calculateG(7.1);

  return Array.from(Array(steps).keys())
    .map((i: number) => i * increments)
    .map((amount: number) =>
      calculate({
        ...parameters,
        surplus: {
          amount,
          include: true,
        },
      })
    )
    .map(
      ({
        afterTotal,
        dividends: { afterDividendsTax, surplus },
        salary: { withoutPension, withoutSurplus },
        tax: { afterTax },
      }) => ({
        total: afterTotal,
        surplus,
        dividends: afterDividendsTax,
        grossIncome: withoutPension,
        netIncome: afterTax,
        salaryBasis: withoutSurplus,
        salaryAboveSevenG: withoutPension > sevenG,
      })
    );
};

export default optimalSurplus;
