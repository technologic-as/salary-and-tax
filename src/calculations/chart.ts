import { OptimalSurplus, optimalSurplus } from './optimal-surplus';
import { defaultSalaryParameters } from './salary';
import { ParamProps } from '../constants';

export type ChartState = {
  total: (any | number)[][];
  dividends: (any | number)[][];
  netIncome: (any | number)[][];
  grossIncome: (any | number)[][];

  maxTotal: any[][];
  surplus: any[][];
  sevenG: number[][];
};

export const getChart = (parameters: ParamProps = defaultSalaryParameters): ChartState => {
  const data = optimalSurplus(parameters);
  const dividends = data.map((i) => [i.surplus, i.dividends]);
  const surplus = data.map((i) => [i.surplus, i.surplus]);
  const grossIncome = data.map((i) => [i.surplus, i.grossIncome]);
  const netIncome = data.map((i) => [i.surplus, i.netIncome]);
  const total = data.map((i) => [i.surplus, i.total]);
  const closeToSeven = data
    .filter((i) => i.salaryAboveSevenG)
    .reduce((a: OptimalSurplus, b) => (a.salaryBasis < b.salaryBasis ? a : b));
  const sevenG = [[closeToSeven.surplus, closeToSeven.grossIncome]];
  const max = data.reduce((a: OptimalSurplus, b) => (a.total > b.total ? a : b));
  const maxTotal = [[max.surplus, max.total]];

  return {
    sevenG,
    maxTotal,
    grossIncome,
    netIncome,
    surplus,
    dividends,
    total,
  };
};

export default getChart;
