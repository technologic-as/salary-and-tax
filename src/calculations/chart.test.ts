import { getChart } from './chart';
import { defaultSalaryParameters } from './index';
import * as optimalSurplus from './optimal-surplus';
import { defaultSalaryParameters as parameters } from './salary';

describe('chart', () => {
  it('should return calculations for a range of different surplus', function () {
    expect(getChart({ ...defaultSalaryParameters, graph: { increments: 100000 } })).toMatchSnapshot();
  });
  it('should get data from optimalSurplus', function () {
    const spy = jest.spyOn(optimalSurplus, 'optimalSurplus');
    getChart(parameters);
    expect(spy).toHaveBeenCalledWith(parameters);
  });
});
