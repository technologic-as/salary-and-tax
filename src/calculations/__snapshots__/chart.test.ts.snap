// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chart should return calculations for a range of different surplus 1`] = `
Object {
  "dividends": Array [
    Array [
      0,
      0,
    ],
    Array [
      100000,
      53290,
    ],
    Array [
      200000,
      106580,
    ],
    Array [
      300000,
      159869,
    ],
    Array [
      400000,
      213159,
    ],
    Array [
      500000,
      266448,
    ],
    Array [
      600000,
      319738,
    ],
    Array [
      700000,
      373028,
    ],
    Array [
      800000,
      426317,
    ],
    Array [
      900000,
      479607,
    ],
    Array [
      1000000,
      532896,
    ],
    Array [
      1100000,
      586186,
    ],
    Array [
      1200000,
      639476,
    ],
    Array [
      1300000,
      692765,
    ],
    Array [
      1400000,
      746055,
    ],
    Array [
      1500000,
      799344,
    ],
    Array [
      1600000,
      852634,
    ],
  ],
  "grossIncome": Array [
    Array [
      0,
      1318257,
    ],
    Array [
      100000,
      1240006,
    ],
    Array [
      200000,
      1162045,
    ],
    Array [
      300000,
      1085358,
    ],
    Array [
      400000,
      1008671,
    ],
    Array [
      500000,
      931984,
    ],
    Array [
      600000,
      855297,
    ],
    Array [
      700000,
      778609,
    ],
    Array [
      800000,
      701923,
    ],
    Array [
      900000,
      625235,
    ],
    Array [
      1000000,
      548548,
    ],
    Array [
      1100000,
      471861,
    ],
    Array [
      1200000,
      395174,
    ],
    Array [
      1300000,
      318487,
    ],
    Array [
      1400000,
      241799,
    ],
    Array [
      1500000,
      165112,
    ],
    Array [
      1600000,
      88192,
    ],
  ],
  "maxTotal": Array [
    Array [
      1400000,
      948171,
    ],
  ],
  "netIncome": Array [
    Array [
      0,
      834690,
    ],
    Array [
      100000,
      792748,
    ],
    Array [
      200000,
      750962,
    ],
    Array [
      300000,
      709857,
    ],
    Array [
      400000,
      668752,
    ],
    Array [
      500000,
      626664,
    ],
    Array [
      600000,
      583258,
    ],
    Array [
      700000,
      539854,
    ],
    Array [
      800000,
      496449,
    ],
    Array [
      900000,
      453043,
    ],
    Array [
      1000000,
      403433,
    ],
    Array [
      1100000,
      353126,
    ],
    Array [
      1200000,
      302818,
    ],
    Array [
      1300000,
      252512,
    ],
    Array [
      1400000,
      202116,
    ],
    Array [
      1500000,
      144036,
    ],
    Array [
      1600000,
      82731,
    ],
  ],
  "sevenG": Array [
    Array [
      700000,
      778609,
    ],
  ],
  "surplus": Array [
    Array [
      0,
      0,
    ],
    Array [
      100000,
      100000,
    ],
    Array [
      200000,
      200000,
    ],
    Array [
      300000,
      300000,
    ],
    Array [
      400000,
      400000,
    ],
    Array [
      500000,
      500000,
    ],
    Array [
      600000,
      600000,
    ],
    Array [
      700000,
      700000,
    ],
    Array [
      800000,
      800000,
    ],
    Array [
      900000,
      900000,
    ],
    Array [
      1000000,
      1000000,
    ],
    Array [
      1100000,
      1100000,
    ],
    Array [
      1200000,
      1200000,
    ],
    Array [
      1300000,
      1300000,
    ],
    Array [
      1400000,
      1400000,
    ],
    Array [
      1500000,
      1500000,
    ],
    Array [
      1600000,
      1600000,
    ],
  ],
  "total": Array [
    Array [
      0,
      834690,
    ],
    Array [
      100000,
      846038,
    ],
    Array [
      200000,
      857542,
    ],
    Array [
      300000,
      869726,
    ],
    Array [
      400000,
      881911,
    ],
    Array [
      500000,
      893112,
    ],
    Array [
      600000,
      902996,
    ],
    Array [
      700000,
      912882,
    ],
    Array [
      800000,
      922766,
    ],
    Array [
      900000,
      932650,
    ],
    Array [
      1000000,
      936329,
    ],
    Array [
      1100000,
      939312,
    ],
    Array [
      1200000,
      942294,
    ],
    Array [
      1300000,
      945277,
    ],
    Array [
      1400000,
      948171,
    ],
    Array [
      1500000,
      943380,
    ],
    Array [
      1600000,
      935365,
    ],
  ],
}
`;
